### ISSUE TITLE
Massive resource use

### ISSUE THREAD

**Issue by tomlut** [RomRider/apexcharts-card#210]
**Checklist**
- [x] I updated the card to the latest version available
- [x] I cleared the cache of my browser
- [x] I verified that I'm really running the lastest version in my browser console
- [x] I checked if there is another issue opened with the same problem

**Describe the bug**
I've been quite impressed with the look and function of apexcharts and was going to replace all my mini graph cards but I have found that apexcharts is far too resource intensive to meet my needs. 

For example, the bar charts in the screenshot below are daily maximums or last values for the past week. There could be thousands of measurements a day from the inverter that need to be processed.

When loading, the charts grind not only the frontend to a halt but the backend too. I get loads of errors that unrelated integrations are taking too long to respond. 

This is with Home Assistant on an i7 mini PC.

I tried setting the graph update to 5 minutes rather than on sensor change. This helped a bit. Except when trying to edit the cards. If the card is still loading data when I try to save it seems to start loading and processing the data again, without cancelling the previous load/process operation. Leading to massive resource use. Closing the page does not help. All the processes continue to run to completion. This pretty much prevents Home Assistant from doing anything else.

I've tried removing the "experimental" colour thresholds. The massive resource use continues to occur. 

The mini-graph card (for all it's other failings) does not do this. I have no idea how or why. Caching perhaps?

**Version of the card**
Version: v1.9.0

**To Reproduce**
This is the configuration I used:
```yaml
type: custom:apexcharts-card
apex_config:
  dataLabels:
    background:
      enabled: false
    style:
      colors:
        - var(--primary-text-color)
graph_span: 7d
header:
  show: true
  title: Daily Energy
series:
  - entity: sensor.energy_inverter_daily
    type: column
    show:
      datalabels: true
    group_by:
      func: max
      duration: 1d
```

**Screenshots**

![Screenshot 2021-09-15 at 10-35-01 Overview - Home Assistant](https://user-images.githubusercontent.com/10679300/133353779-2bfd4af5-0791-40c4-9d28-9401965def03.png)

**Expected behavior**
A bit of optimisation, particularly when editing the card so that it does not bog down the entire system. 

**Desktop (please complete the following information):**
 - Browser Firefox
 - Version 91.0.2

**Smartphone (please complete the following information):**
 - Device: [e.g. iPhoneX]
 - OS: 14.8
 - Browser saffari
 - Version ??

**Additional context**

**Comment by RomRider** [comment]
Caching is disabled while editing the card from the frontend, that's on purpose. Maybe there are some bugs there and I could make it better but I don't use the frontend editor so I never really checked.

With regards to the backend being slow, unfortunately that has nothing to do with the card but to requesting history data (almost exact same code as mini-graph-card). I have an i7 NUC to run home-assistant and I don't experience what you are seeing (and I also load a lot of history data). Once the card is configured, you shouldn't have any issue anymore as caching is enabled again automatically.

With regards to the frontend being slow, all the math is done in frontend but I do not experience that either (neither does a lot of other people using the card). If you hace a slow device for your frontend through, that is quite expected (eg: slow Android tablet, old iPad or equivalent)

**Comment by tomlut** [comment]
>  (almost exact same code as mini-graph-card).

That's weird because I don't see the same behaviour with the min-graph card. I can interrupt it loading to my hearts content when editing via the UI. 

> Once the card is configured, you shouldn't have any issue anymore as caching is enabled again automatically.

As I said, it's better, but if you interrupt the loading/processing it begins again without (apparently) stopping the previous processes. And that's what gets heavy. It starts trying to load all this data 2 or 3 time in parallel. 

>  If you hace a slow device for your frontend through,

No, it's an older but still quite powerful PC.

**Comment by vadipp** [comment]
I have found that the graphs were logging lots of errors in the browser console. 
Seems like this was causing slow loading of the graphs for me.
Disabling adblock for the site removed the errors and improved the performance noticeably.

**Comment by vadipp** [comment]
Sorry, it tunrs out that the errors in the console are just transient. Disabling adblock didn't affect them. The errors just disappeared, and then magically reappeared again. Seems like I'm actually facing bug #17.

**Comment by RomRider** [comment]
@vadipp, you can ignore the errors in the console as those are related to some bugs in apexcharts.js but are not affecting functionalities.

> That's weird because I don't see the same behaviour with the mini-graph card. I can interrupt it loading to my hearts content when editing via the UI.

I don't understand what you mean?
If there is a bug with editing the card in the UI (like having ghost instances of the card still running), it should go away once you close the editor and reload the page.

**Comment by tomlut** [comment]
Still an issue on 2021.11.1

**Comment by sicknesz** [comment]
i have noticed a memory leak when the experimental brush features is enabled that was causing my system to crash when editing any card (frontend client running on a desktop pc intel core i7@4ghz and 8gb of ram, chrome was using all 8gig before crashing the tab, it appears to me that when using template sensors when HA redefined them sometimes the card try to re-render without valid values causing a lots lots of attempt to draw svg path like "M3 %NaN %NaN", i tried debugging but the code it pretty complexe and the culprit seems to be HA sending way too much update statement i had a card that was re-rendering almost every seconds despite the interval set to 10sec, when debugging i could see the interval got reset every time the "hass()" fonction was called ... hope this helps, for the moment i just disabled the brush feature and memory usage is almost back to normal usage (still a bit high but at least stable)

**Comment by tomlut** [comment]
Still an issue on 2021.12.7

**Comment by tomlut** [comment]
Still an issue on 2022.2.0

**Comment by Hamsterman** [comment]
I have an interface with two graphs - its super slow. I wish it would generate the graph serverside as an image so it would be super fast for the client!

**Comment by sicknesz** [comment]
I thinks there's some memory leaks with the card, even with only 2 sensors as inputs, massive CPU/RAM usage, as soon as the card is removed memory goes back to normal usage, and it's even worse when editing (most likely due to cache being disable as you mentionned)

![Screenshot_20220302_191609(1)](https://user-images.githubusercontent.com/3835626/156424039-ca068885-b1af-42d3-915c-7323f3b3204c.png)

**Comment by RomRider** [comment]
There might be some issues with edit, yes and I need to have a look at it. however outside of editing the card, the card is resource heavy especially if you have to display A LOT of datapoints (like using a sensor updated every minute or less with raw data over a long period of time).

**Comment by Hamsterman** [comment]
> There might be some issues with edit, yes and I need to have a look at it. however outside of editing the card, the card is resource heavy especially if you have to display A LOT of datapoints (like using a sensor updated every minute or less with raw data over a long period of time).

I have a sensor with lots of datapoints - what can you recommend me do instead.
Also I have the same data in a SQL database.

**Comment by jazzmonger** [comment]
I thought I was losing my mind... I've chased this for almost a month and finally realized that **Apex charts IS THE CULPRIT**. it started a few core updates ago,  basically the charts are so slow as to render them totally unusable. The Ui freezes completely for anywhere from 10-30 seconds, only on Lovelace tabs that have the charts. 

Bummer.  It was a nice card.  Anyone found a substitute card?

Jeff

**Comment by tomlut** [comment]
The mini graph card seems to handle large data sets better. The graphs aren't as customisable or nice though.

**Comment by zanna-37** [comment]
> I thought I was losing my mind... I've chased this for almost a month and finally realized that **Apex charts IS THE CULPRIT**. it started a few core updates ago, basically the charts are so slow as to render them totally unusable. The Ui freezes completely for anywhere from 10-30 seconds, only on Lovelace tabs that have the charts.
> 
> Bummer. It was a nice card. Anyone found a substitute card? Fwiw, I'm on core-2022.3.7 now, but it's been going on for weeks. Jeff

@jazzmonger, I'm experiencing the same (UI freezes for many seconds and I can't even switch tab in the dashboard until the graph is loaded) and I'm moving toward [Ploty](https://github.com/dbuezas/lovelace-plotly-graph-card). It seems a younger project but it's very powerful, lighting fast, and you can also interact with the graph.
It is not a direct drop-in replacement, but I'm moving to it. The worst issue for me is dbuezas/lovelace-plotly-graph-card#34. But considering the drawbacks of apex, it's nothing.

**Comment by RomRider** [comment]
> Apex charts IS THE CULPRIT

I have a dashboard with 100 charts displayed at once and it's working fine. Just saying there's an issue without providing any data/performance graph isn't going to help with fixing a potential issue.

**Comment by tomlut** [comment]
Yeah but how many data points do each of your graphs use and what features of Apexcharts are you using?

There is definitely an issue using multiple graphs when using sensors that can update every second or so (like from my solar inverter).

**Comment by zanna-37** [comment]
> Yeah but how many data points do each of your graphs use and what features of Apexcharts are you using?
> 
> There is definitely an issue using multiple graphs when using sensors that can update every second or so (like from my solar inverter).

Yeah, me too. I'm visualizing 10h of data with 8/9 traces that update every minute.

**Comment by RomRider** [comment]
Use `update_interval` then, updating every second is way too high.

**Comment by tomlut** [comment]
I wasn't aware of the update interval. I'll give it a go. Yeah I didn't write the SMA Integration that does that.

**Comment by tonyaik** [comment]
With mini-graph you can select points_per_hour, this helps alot with the speed of the graph. 

I have several sensors updating every second, this really slows everything down when trying to use apex. Every page load takes a lot of resources from both backend and browser, even with group_by and func: avg.

I think a solution would be if you specify group_by and func: first, that should only fetch the first value from the database, not all of them and then filter out what to show in the browser. Another solution would be to be able to specify a value on how many data points you want to fetch, similar to points_per_hour.

Also: update_interval isn't working. https://github.com/RomRider/apexcharts-card/issues/293

**Comment by tonyaik** [comment]
For anyone having this issue, I just updated to v1.11.0-dev.1, my resource usage is way down at the moment. I will keep monitoring. But try out v1.11.0-dev.1 if you're having these issues.

**Comment by RomRider** [comment]
For those who don't have `update_interval` working, could you please share your config in https://github.com/RomRider/apexcharts-card/issues/293 as I can't reproduce the issue.

**Comment by RomRider** [comment]
update_interval fixed in 2.0.0

**Comment by convicte** [comment]
The issue go much worse for me in 2.0.0.
It was very slow, taking up to 30 seconds to load on previous versions, but now it's unusable.

My ```update_interval``` is 2 hours and I can't even get the first card to load before it crushes.
I managed to get it loaded after 10 attempts and clicking wait to load.
This is how the graphs should look like
![image](https://user-images.githubusercontent.com/78666770/165482996-0b2e8c2c-17b9-445b-b696-b0fa427feb82.png)
Is there anything I can optimize in my config to improve the situation:
![image](https://user-images.githubusercontent.com/78666770/165483148-a5669da6-7f79-4d38-b85c-4388bcc811ea.png)
![image](https://user-images.githubusercontent.com/78666770/165483864-bb9f1106-79c7-4893-9408-75d8580f08a0.png)

**Comment by tonyaik** [comment]
The issue was very bad for me with 2.0.1 as well. But it magically became better. Try clearing chrome cache etc.

**Comment by convicte** [comment]
Thanks for the offer, but it didn't make difference for me.

**Comment by convicte** [comment]
I was just wondering if there is any action planned for this issue.
While I love the card, since I can't use it, I'll probably have to redo it on another card type to make it work.

Thank you in advance!

**Comment by Heppi911** [comment]
> Use `update_interval` and maybe an aggregate (eg: avg) function over that then to reduce the sample rate to 1min, updating every second is way too high. Every seconds over 10h is 36000 data points, not surprising it is quite slow...

Unfortunately I have to agree with the people complaining about poor performance. The issue appeared with v1.9 , that is why I downgraded to version 1.8.3 to be able to use the brush feature with long (48 h) selectable timeline.
So it is not just a thing of the many data points. Because the number of data points have been the same with the old version. 
Can you remember what ressource hungry feature you added with v1.9? 
Is there a possibility to reduce the datapoints that are loaded for the brush timeline? E.g. loading only every 20th point from Database?
In my opinion avg does not help because ALL values have to be loaded to avg the an then plot as many points as before. Why should it be faster?

generally it would make sense to have the option to display only every xth datapoint. Or even more luxury: To say 3 points per minute. That should be configurable separately for main plot and brush plot.

**Comment by jazzmonger** [comment]
Brutal.... I've had to remove these charts completely.  My inverter current sensors update every 10 seconds. Increasing this value isn't practical as finding max values requires at least 10 sec intervals.  Too bad. This used to be a nice card.

**Comment by pspeirs** [comment]
Hi,

I've been struggling for a week or so on the same issue.  I'm creating a couple of new dashboards using the mushroom templates and finding similar issues to the above comments.  Some of my lovelace editing is done on the UI, however I've noticed that even navigating to a page that contains the charts can start to cause the system to freeze.  I've also noticed it on the back end, most recently while navigating to the backups screen.

It seems that closing down the browser and reopening it fixes the issue until I navigate to a screen that contains a graph.

I am seeing a bunch of errors in the chrome debug console on page load

`apexcharts-card.js?hacstag=3317011521100:1 Error: <svg> attribute height: Expected length, "NaN".
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
setTimeout (async)
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
apexcharts-card.js?hacstag=3317011521100:1 Error: <foreignObject> attribute height: Expected length, "NaN".
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
setTimeout (async)
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
4018Error: <path> attribute d: Expected number, "M 0 NaN".
apexcharts-card.js?hacstag=3317011521100:1 Error: <path> attribute d: Expected number, "M 0 NaN".
attr @ apexcharts-card.js?hacstag=3317011521100:1
plot @ apexcharts-card.js?hacstag=3317011521100:1
path @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
setTimeout (async)
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
apexcharts-card.js?hacstag=3317011521100:1 Error: <path> attribute d: Expected number, "M 0 NaN".
attr @ apexcharts-card.js?hacstag=3317011521100:1
plot @ apexcharts-card.js?hacstag=3317011521100:1
path @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
setTimeout (async)
value @ apexcharts-card.js?hacstag=3317011521100:1
value @ apexcharts-card.js?hacstag=3317011521100:1
(anonymous) @ apexcharts-card.js?hacstag=3317011521100:1
apexcharts-card.js?hacstag=3317011521100:1 Error: <path> attribute d: Expected number, "M 0 NaN".`


In fact, the number of errors goes nuts on each load of the graphs as shown below, over 12000 errors listed.
![image](https://user-images.githubusercontent.com/42822018/173734919-b0ff9a0d-ff9f-465d-a60c-e3db536a2147.png)

Below was a single page load with 3 charts

![image](https://user-images.githubusercontent.com/42822018/173735297-106dba42-ea5d-4a67-b5fe-393a4b68910d.png)

Just removed all the charts, and the system is back to its usual responsive self.

I'm running system on a Pi4 with SSD, system performance params are shown below, and this doesn't really change much.

![image](https://user-images.githubusercontent.com/42822018/173736031-03d7c161-6b0d-4cea-aa26-b3f85eb3b385.png)


I suspect the charts are perhaps affecting the performance of the browser more than anything.

Cheers,
Paul

**Comment by Heppi911** [comment]
@RomRider : Do you see a chance to solve the performance issue within the next 2 month? Otherwise I will definitely have a look on alternatives, because the graphs I setup with lot of time effort cannot be used, when the whole page freezes.

**Comment by pspeirs** [comment]
I would suggest looking for alternatives, there isn't a lot of action happening in this space at the moment.  Remember too that these addins are free, and people do have priorities that change from time to time.

**Comment by klumpo** [comment]
> Brutal.... I've had to remove these charts completely. My inverter current sensors update every 10 seconds. Increasing this value isn't practical as finding max values requires at least 10 sec intervals. Too bad. This used to be a nice card.
> 
> Plotly works flawlessly.

Hi I also have big performance problems with apexcharts-card. The same with plotly. The longer the graph is the slower everything gets.

mini-graph-card however is fine.

I too hope this is improved soon as apexcharts are much prettier and more useful features 😊

**Comment by RomRider** [comment]
> Hi I also have big performance problems with apexcharts-card. The same with plotly. The longer the graph is the slower everything gets.

Please read this comment: https://github.com/RomRider/apexcharts-card/issues/210#issuecomment-1118884848

**Comment by christopherwit** [comment]
As I struggled last evening with bad performance of this otherwise great graph card, I did run into this thread and hopefully help others with my solution. 

Using the option forceNiceScale in the code below resulted in (and got sorted by removing this option again) multiple pop-up messages from Firefox (103.0 64bit) about the hass tab resulting in bad performance of my browser as a whole. Closing these prompts and a good portion of patience, unfreezes my browser. Having a look at the browser debug mode terminal it reported: Uncaught (in promise) out of memory.

For reference: Data plotted in this chart is 2 days of 5 series with an updating each 30min, resulting in roughly 6000 datapoints.

Configuration leading to the issue:
```yaml
type: 'custom:apexcharts-card'
header:
  show: true
  title: Humidity
  show_states: true
  colorize_states: true
apex_config:
  legend:
    show: false
  grid:
    strokeDashArray: 0
    yaxis:
      lines:
        show: false
  yaxis:
    max: 100
    forceNiceScale: true
  stroke:
    width: 2
    curve: smooth
  chart:
    zoom:
      enabled: true
    toolbar:
      show: true
      tools:
        download: false
        zoom: true
        zoomin: true
        zoomout: true
        pan: true
        reset: true
graph_span: 2d
series:
  - entity: sensor.living_room_humidity
  - entity: sensor.babyroom_humidity
  - entity: sensor.bedroom_humidity
  - entity: sensor.outdoor_humidity
  - entity: sensor.metno_liefdesnest_humidity
    show:
      in_header: false
```

**Comment by klumpo** [comment]
> > Hi I also have big performance problems with apexcharts-card. The same with plotly. The longer the graph is the slower everything gets.
> 
> Please read this comment: [#210 (comment)](https://github.com/RomRider/apexcharts-card/issues/210#issuecomment-1118884848)

Hi thanks for the heads up I used  group_by:    func: avg  duration: 20m and it is very fast now, almost as fast as a similar mini graph card.

here is my yaml in case somone else wants it:

`type: custom:apexcharts-card
header:
  show: false
  show_states: true
  colorize_states: true
graph_span: 7d
update_interval: 1min
series:
  - entity: sensor.power_dr_dedichens_vei_48
    stroke_width: 1
    group_by:
      func: avg
      duration: 20m
yaxis:
  - show: false
    min: 0
    max: 10000
    apex_config:
      tickAmount: 3`


`type: custom:mini-graph-card
font_size: 75
line_width: 1
entities:
  - sensor.power_dr_dedichens_vei_48
name: Electricity use
hours_to_show: 168
points_per_hour: 3
show:
  labels: true`

**Comment by mortenalbring** [comment]
Just wanted to confirm that this is still an issue, and I get a lot of console errors on some of my dashboards with Apex charts. 
![image](https://user-images.githubusercontent.com/4549718/183674204-d89a81c0-6bda-4cd0-b159-9ff17cddbbce.png)

However, I can also confirm that using group_by: func: avg makes a _huge_ difference and I don't see any of those console errors when I'm using group_by. I'm guessing there's some spurious or weird data in the series that's getting 'fixed' by the average function? Though I was seeing this on all my dashboards, using Apex charts for a variety of sensors.

**Comment by convicte** [comment]
The issue has not been resolved! 

There are no resolution steps.

**Comment by convicte** [comment]
@RomRider would you mind having another look at it?

Anything that I've tried so far (including all the ideas offered above) do very little.
The graphs are loading dashboard so much that they simply crash or not load at all.
I have to wait anywhere between 10-60 seconds for the different dashboards to load on a PC and nearly for even on a phone.

Below a few of my configs—if you see anything I am doing wrong, I'd greatly appreciate some advice.

```yaml
    cards:
      - type: custom:apexcharts-card
        update_interval: 30min
        experimental:
          brush: true
        graph_span: 7d
        brush:
          selection_span: 72h
        header:
          show: true
          title: Solar power managment - daily
          show_states: true
          colorize_states: true
        yaxis:
          - max: '|+20|'
        series:
          - entity: sensor.daily_solar_energy_production
            name: Daily Solar production
            type: column
            group_by:
              func: median
              duration: 60min
            show:
              in_brush: true
          - entity: sensor.daily_energy_consumed
            name: Daily Incoming energy
            group_by:
              func: median
              duration: 60min
          - entity: sensor.daily_energy_returned
            name: Daily Returned energy
            group_by:
              func: median
              duration: 60min
          - entity: sensor.absolute_energy_consumption
            name: Daily Absolute energy consumption
            group_by:
              func: median
              duration: 60min
  ```
  and
  
  ````yaml
  
        - type: custom:apexcharts-card
        update_interval: 10min
        experimental:
          brush: true
        graph_span: 7d
        brush:
          selection_span: 48h
        header:
          show: true
          title: Power flow
          show_states: true
          colorize_states: true
        series:
          - entity: sensor.net_power_meter
            type: column
            group_by:
              func: avg
              duration: 30min
            show:
              in_brush: true
          - entity: sensor.solar_ac_power
            type: line
            stroke_width: 1
            name: Current solar output
            group_by:
              func: avg
              duration: 30min
            show:
              in_brush: true
          - entity: sensor.returned_power_meter
            type: column
            group_by:
              func: avg
              duration: 30min
            show:
              in_brush: true
          - entity: sensor.incoming_power_meter
            type: column
            group_by:
              func: avg
              duration: 30min
            show:
              in_brush: true

      - type: custom:apexcharts-card
        update_interval: 10min
        experimental:
          brush: true
        graph_span: 7d
        brush:
          selection_span: 72h
        header:
          show: true
          title: Solar power managment - momentary
          show_states: true
          colorize_states: true
        series:
          - entity: sensor.yearly_solar_energy_production
            name: Momentary Solar production
            type: column
            group_by:
              func: diff
              duration: 30min
            show:
              in_brush: true
          - entity: sensor.yearly_energy_consumed
            name: Momentary Incoming energy
            group_by:
              func: diff
              duration: 30min
          - entity: sensor.yearly_energy_returned
            name: Momentary Returned energy
            group_by:
              func: diff
              duration: 30min

**Comment by convicte** [comment]
Example of a failure to load, but there are many other ways it fails:

![image](https://user-images.githubusercontent.com/78666770/194648734-edacd20e-ffbe-4bde-995a-91fdb4dc92c2.png)

**Comment by mcfly2283** [comment]
I have the sam Issue. I Can`t use the card with my Android-Wallpannel anymore, due to freezing the UI completely for about 5 Seconds while the Card is reloading. On an i7 it`s take about 2 Seconds.

There are only 2 Sensors without many Data-Points. 

      - type: custom:apexcharts-card
        graph_span: 24h
        experimental:
          disable_config_validation: true
        all_series_config:
          stroke_width: 2
          fill_raw: last
        header:
          show: false
          show_states: false
          colorize_states: true
        series:
          - entity: sensor.0x158d00044d1e69_humidity
            name: Haus
            type: area
            show:
              in_header: false
              legend_value: true
          - entity: sensor.delta_forecast_erdgeschoss
            name: Forecast
            show:
              in_header: false
              legend_value: true
            unit: '%'
            type: line

![image](https://user-images.githubusercontent.com/34608100/196341067-2ddac842-1c38-4664-a799-b6c30527443f.png)

**Comment by philsson** [comment]
I second this bug causing both backend and frontend to freeze.

**Comment by alphasixtyfive** [comment]
I can confirm having same issue.

**Comment by convicte** [comment]
Thanks for the prompt comment. I agree with mostly all of what you said.

1) Plotly vs. APEX - agreed, APEX is way more polished
2) I am mostly using ``group_by`` - please see https://github.com/RomRider/apexcharts-card/issues/210#issuecomment-1272066956 - and it makes a big difference but after having it added it's still mighty slow.
3) Since there are that many people reproducing it, if there would be will to tackle it, I am guessing there would be a line of people wanting to provide logs, input, and beta test. I am not sure if I subscribe to the 'difficult to reproduce' being the root cause.

**Comment by pspeirs** [comment]
Remember too that if someone else wishes to have a crack at it, fork the project.

**Comment by convicte** [comment]
@PierreMarieBaty respectfully that is the most TL:DR while mansplaining response I've seen in a while.

No, that is not a fix (not even a workaround) nor is it any different that majority of the people in the thread reported to have been using from the start and still experiencing performance issues.

Not to mention that averaging is only possible with continuous variables while at least some people reported having worked with entities that cannot be averaged at all.

So TL:DR - it's always a good idea to at least read the last 5 comments before making such authoritative statements.

**Comment by PierreMarieBaty** [comment]
@convicte Apologies for the mansplaining. The TL;DR diagnosis is right though, I totally endorse it. I wrote that from the point of view of the average user looking for a usable multichart card for HomeAssistant who doesn't expect to spend hours on it. Pretty sure I'm not the only one : without that setting even the provided graph samples are unusable. As there was no indication that feeding a large dataset to the graph code would freeze the UI, and considering the average user was not supposed to expect that behavior as competing chart solutions handle them well, I was about to throw it away and pester against yet another open source garbage, but gave it a second thought and attempted to "fix" that myself. Call that as you wish, not a fix not even a workaround, it makes the solution usable. That's all the average user needs to know, and I persist thinking it should be a default setting.

**Comment by Ferrox85** [comment]
I'm having the same issue. It freezes the complete HASS system on both ends. Console gives the following errors when trying to reload the graph:

<img width="2040" alt="Scherm­afbeelding 2023-01-16 om 10 53 34" src="https://user-images.githubusercontent.com/64791738/212649439-a1ad6ade-4529-427f-801a-fe5fd747fb34.png">

The code I use:

type: custom:apexcharts-card
graph_span: 10d
update_interval: 1h
stacked: true
header:
  show: true
  title: Opbrengst PV
yaxis:
  - id: first
  - id: second
    opposite: true
series:
  - entity: sensor.opbrengst_zonnepanelen
    name: Omzet PV
    type: column
    yaxis_id: first
    opacity: 0.8
    color: '#03b2cb'
    group_by:
      duration: 1d
      func: diff
    show:
      legend_value: false
  - entity: sensor.netto_kosten_zonnepanelen
    name: ROI PV
    yaxis_id: second
    type: line
    opacity: 0.8
    color: '#cb6f02'
    group_by:
      duration: 1d
      func: last
    show:
      legend_value: false

It doesn't load longer than 10days in my experience. I assume it uses the short term statistics, where it would be better to use long term statistics with this setup. The data is updated every 5 seconds. So it generates a lot of datapoints. 

I can read code, but haven't developed for HASS yet. Hence, I am using HASS for a few months now. I want more of these type of graphs, because they are very beautiful in giving insight in daily volumes (ex. power usage an total usage) If I wan't to debug this I need time and soms hints where to start.

**Comment by mortenalbring** [comment]
> I'm having the same issue. It freezes the complete HASS system on both ends. Console gives the following errors when trying to reload the graph:

Is it possible to use 

    group_by:
      func: avg
      duration: 1h

in your case? As far as I'm aware, that's still the only known workaround for this issue. It's annoying to have to add it in to _every_ chart every time, but it does seem to solve the issue for most of us.

**Comment by Ferrox85** [comment]
I can do that, but I want the diff between the days. I can confirm it is a bug in backend of HASS. Because other front end implementations suffer the same slow performance. (like https://github.com/dbuezas/lovelace-plotly-graph-card) for instance. It is possible to achieve the same result but does take a long time either to load. Not shure what to doe next. rebuidling my database as we speak but I don't think that is going to be the solution.

**Comment by roblomq** [comment]
For those who are intereseted, I had the same experience with the extreme slow performance after I added a apex chart to my HA. It took me verry long to be even able to remove the chart. Seems that it does not work well for me so I removed apex chart.

**Comment by alphasixtyfive** [comment]
> For those who are intereseted, I had the same experience with the extreme slow performance after I added a apex chart to my HA. It took me verry long to be even able to remove the chart. Seems that it does not work well for me so I removed apex chart.

I have same expirience, it takes up  to 3 mins for some charts to load yet there is no alternative so far which have time on X scale sadly.
